### PR TITLE
[DQT] Downloadable Fields

### PR DESCRIPTION
**Pull request by jstirling91**
check to see if fields are files when loading saved query

### CODE REVIEW

File: modules/dataquery/jsx/react.app.js [hunk 1]
```diff
@@ -440,6 +440,26 @@ var DataQueryApp = React.createClass({
...
         });
+        for (var i = 0; i < fieldsList.length; i++) {
+            var that = this;
+            $.ajax({
+                url: loris.BaseURL + "/AjaxHelper.php?Module=dataquery&script=datadictionary.php",
+                success: function(data) {
+                    if(data[0].value.IsFile) {
+                        that.setState(function(state){
+                            var key = data[0].key[0] + "," + data[0].key[1]
+                            var downloadable = state.downloadableFields;
```
Comment (jacobpenny) [inline modules/dataquery/jsx/react.app.js]: These two variables can be `const`

File: modules/dataquery/jsx/react.app.js [hunk 2]
```diff
@@ -440,6 +440,26 @@ var DataQueryApp = React.createClass({
                 loading: false
             }
         });
+        for (var i = 0; i < fieldsList.length; i++) {
+            var that = this;
+            $.ajax({
+                url: loris.BaseURL + "/AjaxHelper.php?Module=dataquery&script=datadictionary.php",
+                success: function(data) {
```
Comment (jacobpenny) [inline modules/dataquery/jsx/react.app.js]: If you use an arrow function here, i.e. ```(data) => { ... } ``` you can remove the above `var that = this;`. Arrow function don't bind their own `this`.

File: modules/dataquery/jsx/react.app.js [hunk 3]
```diff
@@ -440,6 +440,26 @@ var DataQueryApp = React.createClass({
                 loading: false
             }
         });
+        for (var i = 0; i < fieldsList.length; i++) {
+            var that = this;
```
Comment (jstirling91) [inline modules/dataquery/jsx/react.app.js]: @jacobpenny Yes this can but this code is not using es6 syntax and is staying consistent with the format of the existing code

File: modules/dataquery/jsx/react.app.js [hunk 4]
```diff
@@ -440,6 +440,26 @@ var DataQueryApp = React.createClass({
                 loading: false
             }
         });
+        for (var i = 0; i < fieldsList.length; i++) {
```
Comment (jstirling91) [inline modules/dataquery/jsx/react.app.js]: @jacobpenny I would agree that `foreach` would be a better choice if the code within the block was syncronise, but since all the loop does is make async calls it seems to be an added layer of complexity that is not needed

**Comment by jstirling91** [comment]
@jacobpenny I think for this pull request sticking with es5 syntax is best. All the code is written in it so its best to stay consist. Though in another pull request I can update the syntax to become more consistent with the loris javascript standards

File: modules/dataquery/jsx/react.app.js (hunk 3 above)
Comment (jacobpenny) [inline modules/dataquery/jsx/react.app.js]: Fair enough

**Comment by driusan** [comment]
Why did you remove the Needs Rebase tag?

It's still showing up as having conflicts in `modules/dataquery/js/react.app.js`

**Comment by driusan** [comment]
nevermind, I just realized that was 19 days ago.. after you've rebased it, go ahead and move it to Senior Review

File: modules/dataquery/jsx/react.app.js [hunk 5]
```diff
@@ -466,6 +466,26 @@ var DataQueryApp = React.createClass({
...
+            var that = this;
+            $.ajax({
+                url: loris.BaseURL + "/AjaxHelper.php?Module=dataquery&script=datadictionary.php",
```
Comment (driusan) [inline modules/dataquery/jsx/react.app.js]: This URL should be `loris.BaseURL + "/dataquery/ajax/datadictionary.php"`, not directly calling `AjaxHelper`

File: modules/dataquery/ajax/datadictionary.php [hunk 6]
```diff
@@ -19,16 +19,34 @@
...
+} else if ($_REQUEST['key']) {
+    $key = explode(',', $_REQUEST['key']);
+    // error_log($key);
+
+    $results = $cdb->queryView(
+        "DQG-2.0",
+        "datadictionary",
+        array(
+         "reduce" => "false",
+         "key"    => "[\"$key[0]\",\"$key[1]\"]",
```
Comment (driusan) [inline modules/dataquery/ajax/datadictionary.php]: I don't think this is any worse than before, but shouldn't these keys be `urlencode()`ed? Or does queryView do that already? (same with the $category variable above..)

Comment (jstirling91) [inline modules/dataquery/ajax/datadictionary.php]: If its to be URL encoded we should always do it in the backend side since users could make their own request. I'll add it, though we should really create a more uniform way of getting `$_REQUEST` variables in a future release of LORIS and enforce all request variables to use that

File: modules/dataquery/ajax/datadictionary.php [hunk 7]
```diff
@@ -19,16 +19,33 @@
...
+        "DQG-2.0",
+        "datadictionary",
+        array(
+         "reduce"   => "false",
+         "startkey" => "[\"$category\"]",
+         "endkey"   => "[\"$category\", \"ZZZZZZZZ\"]",
+        )
+    );
+} else if ($_REQUEST['key']) {
+    $key = explode('%2C', urlencode($_REQUEST['key']));
```
Comment (jstirling91) [inline modules/dataquery/ajax/datadictionary.php]: @driusan I've done the encoding here

**Comment by driusan** [comment]
needs rebase again.. (I think because the "null filters" PR was merged)

**Comment by driusan** [comment]
It doesn't look like this was rebased properly (GitHub shows 103 files changed now, with most of them looking unrelated to this PR..)

**Comment by jstirling91** [comment]
I hate github

File: modules/dataquery/ajax/datadictionary.php [hunk 8]
```diff
@@ -19,16 +19,33 @@
...
+if ($_REQUEST['category']) {
+    $category = urlencode($_REQUEST['category']);
+
+    $results = $cdb->queryView(
+        "DQG-2.0",
+        "datadictionary",
+        array(
+         "reduce"   => "false",
+         "startkey" => "[\"$category\"]",
+         "endkey"   => "[\"$category\", \"ZZZZZZZZ\"]",
```
Comment (xlecours) [inline modules/dataquery/ajax/datadictionary.php]: "endkey"   => "[\"$category\", \"{}\"]", or "endkey"   => "[\"$category\", \"\ufff0\"]"
Objects are ordered last in the view indexes. ufff0 would realy be the last value if we need non-empty object. (But I doubt it)

Comment (jstirling91) [inline modules/dataquery/ajax/datadictionary.php]: I'm not quite sure, this is just legacy code that I didnt touch just indented. If this is the case I think this should be added to a separate pull request.
